sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -66,11 +66,12 @@ case class PrintToStderr(child: Expression) extends UnaryExpression {
   """,
   since = "3.1.0",
   group = "misc_funcs")
-case class RaiseError(child: Expression) extends UnaryExpression with ImplicitCastInputTypes {
+case class RaiseError private[spark] (child: Expression, returnType: DataType)
+  extends UnaryExpression with ImplicitCastInputTypes {
 
   override def foldable: Boolean = false
   override def nullable: Boolean = true
-  override def dataType: DataType = NullType
+  override def dataType: DataType = returnType
```

Comment:
nit: the constructor parameter can be `dataType` directly, then we don't need this override.

Reply:
sgtm， updated